### ISSUE TITLE
Can we mask hole image object using this

### ISSUE THREAD

**Issue by thakur0703** [akirasosa/mobile-semantic-segmentation#26]

**Comment by ldenoue** [comment]
can you be specific?
I would also like to have both hair and face in my masks, but I only can get either hair or face.

**Comment by thakur0703** [comment]
Particular object mask within a image.

**Comment by ldenoue** [comment]
with Line #26 you can only pick one channel: 0 for hair, 1 for face and 2 for background.
To merge several channels into one mask, you have to extract each channel and sum them.

**Comment by thakur0703** [comment]
On which class?

**Comment by ldenoue** [comment]
in this line https://github.com/akirasosa/mobile-semantic-segmentation/blob/master/data.py#L28
you can change 0 (hair) to 1 (face) mask_iter = mask_gen.flow(np.expand_dims(masks[:, :, :, **0**], axis=4) to
mask_iter = mask_gen.flow(np.expand_dims(masks[:, :, :, **1**], axis=4)

**Comment by jbmaxwell** [comment]
I'm trying to convert this model to handle multiple classes (https://github.com/akirasosa/mobile-semantic-segmentation/issues/31). As a first step, I've got it running on all three classes by changing the last Conv2D layer to `Conv2D(3, (1,1) ...)(b18)`, and by removing `np.expand_dims(masks[:, :, :, 0], axis=4)` entirely and just passing in `masks` (i.e., all three channels). However, I have a feeling this changes the loss function somehow, as only one epoch (which was much slower, of course) seems to have surprisingly good results: 
`loss: -0.7867 - dice_coef: 0.7867 - recall: 0.8388 - precision: 0.7576 - binary_crossentropy: 0.8161 - val_loss: -0.8347 - val_dice_coef: 0.8347 - val_recall: 0.8314 - val_precision: 0.8424 - val_binary_crossentropy: 0.6801`
Mind you, I suppose this may be because it can be less discriminating (i.e., it's not filtering out of the image)?

**Comment by jbmaxwell** [comment]
In case anyone's interested, in future; to get proper (or at least more conventional) multi-class segmentation working I basically followed the lead in this post at keras-team: https://github.com/keras-team/keras/issues/1169. 
As mentioned above, I had to let `_create_datagen` (datay.py) send the original mask: `mask_iter = mask_gen.flow(masks, seed=seed)`, and also reshape the return of the iterator: `yield img, np.reshape(mask, (mask.shape[0], int(mask.shape[1])*int(mask.shape[2]), mask.shape[3]))`. 
Then the bottom of my `MobileUNet` is similarly reshaped, and changed to sigmoid: 
`
    x = Reshape((int(x.shape[1])*int(x.shape[2]), num_classes))(x)
    x = Activation('softmax')(x)
`
That seems to work with the provided data. For my own data, which is a more conventional integer labelled 2D image format, I'll have to expand the pixel labels out to binary channels... Not sure how to do that, so if anyone has any tips they'd be appreciated.

**Comment by akirasosa** [comment]
If I have time, I will put an example of multiclass segmentation such as hair, face and background. Thanks,
